### DISCUSSION TITLE
Lesson 7: function endLottery not working

### DISCUSSION THREAD

**Discussion by Karlus44** [smartcontractkit/full-blockchain-solidity-course-py#1265] [Q&A] [answered]
Hello there, I encontered a problem when I test my project, any help would be greatly appreciated because I've been stuck here for weeks now.
All my code can be found in here: https://github.com/Karlus44/smartcontract-lottery
When I type the command brownie test
my script tests/test_lottery_unit.py is exectuted, and my different tests are submitted.
The problem seems to be in my endLottery function, and more specifically during the call requestRandomness, and then LINK.transferAndCall.
I don't see what I've done wrong.
Here some quotes of my logout. This is on my development network, but the same happens in my "ganache-local" and also on rinkeby:

> platform linux -- Python 3.8.10, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /home/karlus/Documents/NFT-Contract/smartcontract-lottery
plugins: eth-brownie-1.18.1, forked-1.4.0, web3-5.27.0, hypothesis-6.27.3, xdist-1.34.0
collected 4 items

> 
> Launching 'ganache-cli --accounts 10 --hardfork istanbul --gasLimit 12000000 --mnemonic brownie --port 8545'...
> 
> tests/test_lottery_unit.py ...F                                                    [100%]
> 
>    def test_can_end_lottery():
>         # #
>         lottery.enter({"from": account, "value":lottery.getEntranceFee()})
>         fund_with_link(lottery)
>         print(f"endlottery: {lottery.endLottery}")
>       transaction = lottery.endLottery({"from": account})
> E       brownie.exceptions.VirtualMachineError: revert
> E       Trace step -1, program counter 2469:
> E         File /smartcontractkit/chainlink-brownie-contracts@1.1.1/contracts/src/v0.6/VRFConsumerBase.sol", line 161, in 
> VRFConsumerBase.requestRandomness:
> E           function requestRandomness(bytes32 _keyHash, uint256 _fee)
> E             internal returns (bytes32 requestId)
> E           {
> E             LINK.transferAndCall(vrfCoordinator, _fee, abi.encode( _keyHash, USER_SEED_PLACEHOLDER));
> E             // This is the seed passed to VRFCoordinator. The oracle will mix this with
> E             // the hash of the block containing this request to obtain the seed/input
> E             // which is finally passed to the VRF cryptographic machinery.
> 
> tests/test_lottery_unit.py:63: VirtualMachineError

> ----------------------- Captured stdout call -----------------------
0x66aB6D9362d4F35596279692F0251Db635165871
Deployed lottery!
Fund contract!
endlottery: <ContractTx 'endLottery()'>
Transaction sent: 0xaf6a01c8d6d0b18fc410b4b0c3c882da8dd4158e8871d2a73a4326e3d636d55d
===================== short test summary info ======================
FAILED tests/test_lottery_unit.py::test_can_pick_winner_correctly
============== 1 failed, 3 passed, 1 skipped in 8.25s ==============
Terminating local RPC client...

**Comment by n4n0b1t3** [comment]
Hi @Karlus44 I am at the same lesson. I would try to help you to find out what doesn't work but the setup makes it very hard for me.
1. Using the wrong format for your terminal output makes my eyes bleed. Have a look at this post here: https://github.com/smartcontractkit/full-blockchain-solidity-course-py/discussions/1176
2. The VirtualMachineError looks for me like the transaction was reverted. This can be tested in remix, did you try this?
3. Your repo looks very similar to the original one here: https://github.com/PatrickAlphaC/smartcontract-lottery I don't know who would have the time to comb through it to find the difference between both. It would make it easier for others to help you if you could point out, where your code differs.

**Comment by Karlus44** [comment]
Hi @n4n0b1t3 thanks for willing to help.
I edited the log and hope it will be more easy to read like this.
On rinkeby, I have something like that:

>         fund_with_link(lottery)
>        lottery.endLottery({"from": account, "gas_limit": 6721975, "allow_revert": True})
>  E       brownie.exceptions.RPCRequestError: Accessing `TransactionReceipt.revert_msg` on a reverted transaction requires the `debug_traceTransaction` RPC endpoint, but the node client does not support it or has not made it available.
> 
> tests/test_lottery_integration.py:16: RPCRequestError
> ----------------------- Captured stdout call -----------------------
> 0xe53D811C533AE8161e45a129f580c3f658681DF4
>
> Waiting for https://api-rinkeby.etherscan.io/api to process contract
> ...
> Verification submitted successfully. Waiting for result...
> Verification complete. Result: Already Verified
> Deployed lottery!
> Fund contract!
> ===================== short test summary info ======================
> FAILED tests/test_lottery_integration.py::test_can_pick_winner - ...

So a different message, but it seems to be the same error. At least it's the same function that fails.

I'm trying to compile my contract on remix, but I don't know how to use the right version of VRFConsumerBase and OpenZeppelin.

I think my code doesn't differ from the original. I just added some print to see what happens but I believe the problem is not in the code but could be in some network settings ? (I could be wrong of course)

**Reply by n4n0b1t3** [reply]
I won't be of much help solving your problem. But I would suggest using notepad++ (if you are on Windows) with the compare plugin. This allows you to open two files and it will mark the differences. The Left and right panes will scroll simultaniously. You can then compare your version and the original repository. In my case, my version does not compile but the original does, now I will remove piece by piece all alterations from my personal version until it compiles. This way I can find out, where the problem is. I wrote an article about manual unaided debugging, this might help to get into the right spirit: https://n4n0b1t3.medium.com/manual-and-unaided-debugging-of-solidity-python-blockchain-applications-a2cf7a1049be

**Reply by n4n0b1t3** [reply]
> I think my code doesn't differ from the original. 

Do you use quotes for addresses on .env or on brownie-config.yaml?
.env doesn't need quotes and the example files contain wrong formats with backticks mixed into it.
.yaml needs single quotes.
Just a thought...

**Comment by n4n0b1t3** [comment, marked as answer]
Did you try this: https://github.com/smartcontractkit/full-blockchain-solidity-course-py/discussions/1267

**Comment by Karlus44** [comment]
ok @n4n0b1t3 thanks for your valuable help, I did solve my problem.
First I did download and test the code from PatrickAlpha, and it worked fine, so I could see that the problem was in my code after all.
I'm not using Windows, so I didn't try Notepad. Maybe there is actually notepad in Linux, or maybe my editor Atom already allows the functionality of comparing texts. Anyway, before I get to this I tried to compare the files with my bare eyes and saw that I was giving a wrong fee amount in my brownie-config.yaml file. 
Although I believed I was giving the right amount here:
`def fund_with_link(contract_address, account=None, link_token=None, amount=10**17):  # 0.1 Link`,
the fee was actually read in this file.
Well, thank you again !

**Reply by n4n0b1t3** [reply]
Thx, glad it helped!
